### PR TITLE
samba4: add wrapPythonPrograms

### LINKED ISSUE #132355: samba-tool crashes (missing dependencies)
### Describe the bug
`samba-tool` is an integral tool for creating, managing and backing up Samba domain controllers. It is only built in the samba4Full package when DC support is enabled. Simply running `samba-tool` is enough to trigger a crash since the Python dependency `tdb` is not present.

### Steps To Reproduce
1. `nix-shell -p samba4Full --command samba-tool`

### Expected behavior
`samba-tool` displays a help message.

### Screenshots
```console
ERROR(<class 'ModuleNotFoundError'>): uncaught exception - No module named 'tdb'
  File "/nix/store/5ksfz6ba9c3g03sczapdcr3c2ak3acbh-samba-4.14.4/bin/samba-tool", line 44, in <module>
    retval = cmd._run("samba-tool", subcommand, *args)
  File "/nix/store/5ksfz6ba9c3g03sczapdcr3c2ak3acbh-samba-4.14.4/lib/python3.8/site-packages/samba/netcmd/__init__.py", line 259, in _run
    cmd = self.subcommands[cmd_name]
  File "/nix/store/5ksfz6ba9c3g03sczapdcr3c2ak3acbh-samba-4.14.4/lib/python3.8/site-packages/samba/netcmd/main.py", line 35, in __getitem__
    self[attr] = getattr(__import__('samba.netcmd.%s' % package,
  File "/nix/store/5ksfz6ba9c3g03sczapdcr3c2ak3acbh-samba-4.14.4/lib/python3.8/site-packages/samba/netcmd/delegation.py", line 23, in <module>
    from samba import provision
  File "/nix/store/5ksfz6ba9c3g03sczapdcr3c2ak3acbh-samba-4.14.4/lib/python3.8/site-packages/samba/provision/__init__.py", line 50, in <module>
    from samba.samba3 import smbd, passdb
  File "/nix/store/5ksfz6ba9c3g03sczapdcr3c2ak3acbh-samba-4.14.4/lib/python3.8/site-packages/samba/samba3/__init__.py", line 27, in <module>
    import tdb
```

### Additional context
* A workaround exists [here - #105486 (comment)](https://github.com/NixOS/nixpkgs/pull/105486#issuecomment-748497131).
* A pull request that fixes this was submitted [here - #105486](https://github.com/NixOS/nixpkgs/pull/108177), but was withdrawn without any additional information.

### Notify maintainers
@aneeshusa

### Metadata
```console
[user@host:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"aarch64-linux"`
 - host os: `Linux 5.10.17, NixOS, 21.05.1669.973910f5c31 (Okapi)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.14`
 - channels(root): `"nixos-21.05.1669.973910f5c31"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

```

### PR DESCRIPTION
**Pull request by Artturin**
###### Motivation for this change
Closes #132355
Supersedes #108177 Thanks @dasJ for the original pr

###### Things done
Currently compiling

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 132373` at f49c3bb7 run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package marked as broken and skipped:</summary>
  <ul>
    <li>libvirt_5_9_0</li>
  </ul>
</details>
<details>
  <summary>191 packages skipped due to time constraints:</summary>
  <ul>
    <li>MIDIVisualizer</li>
    <li>adapta-gtk-theme</li>
    <li>almanah</li>
    <li>arcan.arcan</li>
    <li>arcan.arcan-wrapped</li>
    <li>arcan.durden-wrapped</li>
    <li>arcan.everyone-wrapped</li>
    <li>arcan.pipeworld-wrapped</li>
    <li>arcan.prio-wrapped</li>
    <li>arcan.xarcan</li>
    <li>...</li>
  </ul>
</details>
<details>
  <summary>32 packages built successfully:</summary>
  <ul>
    <li>ffmpeg-full (arcan.ffmpeg)</li>
    <li>easysnap</li>
    <li>enum4linux</li>
    <li>enum4linux-ng</li>
    <li>evolution-data-server (gnome.evolution-data-server ,gnome.evolution_data_server)</li>
    <li>gnome-online-accounts (gnome.gnome-online-accounts ,gnome.gnome_online_accounts)</li>
    <li>gnome.cheese</li>
    <li>gnome.gnome-session (gnome.gnome_session)</li>
    <li>gnome.gnome-shell (gnome.gnome_shell)</li>
    <li>gnome.gnome-user-share</li>
    <li>grilo-plugins (gnome.grilo-plugins)</li>
    <li>libgdata (gnome.libgdata)</li>
    <li>gnome.nautilus</li>
    <li>gnome.totem</li>
    <li>tracker-miners (gnome.tracker-miners)</li>
    <li>gvfs (gnome2.gvfs)</li>
    <li>imagination</li>
    <li>libvirt</li>
    <li>php74Extensions.smbclient</li>
    <li>php80Extensions.smbclient</li>
    <li>python38Packages.pyglet</li>
    <li>python38Packages.pysmbc</li>
    <li>python39Packages.pydub</li>
    <li>python39Packages.pyglet</li>
    <li>python39Packages.pysmbc</li>
    <li>rabbitvcs</li>
    <li>restream</li>
    <li>samba (samba4)</li>
    <li>sanoid</li>
    <li>smbnetfs</li>
    <li>zfs (zfsStable ,zfsUnstable)</li>
    <li>zfstools</li>
  </ul>
</details>
<details>
  <summary>7 suggestions:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-patch-comment.md">missing-patch-comment</a>

Consider adding a comment explaining the purpose of this patch on the line preceeding.
Near pkgs/servers/samba/4.x.nix:60:5:
```
   |
60 |     ./patch-source3__libads__kerberos_keytab.c.patch
   |     ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/unclear-gpl.md">unclear-gpl</a>

`gpl3` is a deprecated license, please check if project uses `gpl3Plus` or `gpl3Only` and change `meta.license` accordingly.

Near pkgs/servers/samba/4.x.nix:186:5:
```
    |
186 |     license = licenses.gpl3;
    |     ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-patch-comment.md">missing-patch-comment</a>

Consider adding a comment explaining the purpose of this patch on the line preceeding.
Near pkgs/servers/samba/4.x.nix:59:5:
```
   |
59 |     ./4.x-no-persistent-install.patch
   |     ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-patch-comment.md">missing-patch-comment</a>

Consider adding a comment explaining the purpose of this patch on the line preceeding.
Near pkgs/servers/samba/4.x.nix:63:5:
```
   |
63 |     ./build-find-pre-built-heimdal-build-tools-in-case-of-.patch
   |     ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-patch-comment.md">missing-patch-comment</a>

Consider adding a comment explaining the purpose of this patch on the line preceeding.
Near pkgs/servers/samba/4.x.nix:61:5:
```
   |
61 |     ./4.x-no-persistent-install-dynconfig.patch
   |     ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/build-tools-in-build-inputs.md">build-tools-in-build-inputs</a>

python3.pkgs.wrapPython is a build tool so it likely goes to `nativeBuildInputs`, not `buildInputs`.

Near pkgs/servers/samba/4.x.nix:85:3:
```
   |
85 |   buildInputs = [
   |   ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-patch-comment.md">missing-patch-comment</a>

Consider adding a comment explaining the purpose of this patch on the line preceeding.
Near pkgs/servers/samba/4.x.nix:62:5:
```
   |
62 |     ./4.x-fix-makeflags-parsing.patch
   |     ^
```
</li>
  </ul>
</details>

**Comment by Artturin** [comment]
```
/result/bin/samba-tool
Usage: samba-tool <subcommand>

Main samba administration tool.


Options:
  -h, --help       show this help message and exit

  Version Options:
    -V, --version  Display version number


Available subcommands:
  computer    - Computer management.
  contact     - Contact management.
  dbcheck     - Check local AD database for errors.
  delegation  - Delegation management.
  dns         - Domain Name Service (DNS) management.
  domain      - Domain management.
  drs         - Directory Replication Services (DRS) management.
  dsacl       - DS ACLs manipulation.
  forest      - Forest management.
  fsmo        - Flexible Single Master Operations (FSMO) roles management.
  gpo         - Group Policy Object (GPO) management.
  group       - Group management.
  ldapcmp     - Compare two ldap databases.
  ntacl       - NT ACLs manipulation.
  ou          - Organizational Units (OU) management.
  processes   - List processes (to aid debugging on systems without setproctitle).
  rodc        - Read-Only Domain Controller (RODC) management.
  schema      - Schema querying and management.
  sites       - Sites management.
  spn         - Service Principal Name (SPN) management.
  testparm    - Syntax check the configuration file.
  time        - Retrieve the time on a server.
  user        - User management.
  visualize   - Produces graphical representations of Samba network state.
For more help on a specific subcommand, please type: samba-tool <subcommand> (-h|--help)
```

```
./result/bin/samba_dnsupdate
The server update list was not found, and --update-list was not provided.
[Errno 2] No such file or directory: '/var/lib/samba/private/dns_update_list'

Usage: samba_dnsupdate [options]
```

**Review by dasJ — COMMENTED** [review]
I would have opted for `python3.pkgs` instead of `python3Packages` so I can override both Python and the packages with a single override, e.g. `samba4.override { python3 = python39; }`

**Comment by Artturin** [comment]
> I would have opted for `python3.pkgs` instead of `python3Packages` so I can override both Python and the packages with a single override, e.g. `samba4.override { python3 = python39; }`

switched all to python3Packages

File: pkgs/servers/samba/4.x.nix [hunk 1]
```diff
@@ -64,7 +63,7 @@ stdenv.mkDerivation rec {
   ];
 
   nativeBuildInputs = [
-    python3
+    python3Packages.python
```
Comment (SuperSandro2000) [inline pkgs/servers/samba/4.x.nix:66]:
```suggestion
    python3
```

File: pkgs/servers/samba/4.x.nix [hunk 2]
```diff
@@ -83,7 +82,8 @@ stdenv.mkDerivation rec {
   ];
 
   buildInputs = [
-    python3
+    python3Packages.python
+    python3Packages.wrapPython
```
Comment (SuperSandro2000) [inline pkgs/servers/samba/4.x.nix:86]:
```suggestion
    python3
    python3.pkgs.wrapPython
```

File: pkgs/servers/samba/4.x.nix [hunk 3]
```diff
@@ -146,6 +146,8 @@ stdenv.mkDerivation rec {
   # module, which works correctly in all cases.
   PYTHON_CONFIG = "/invalid";
 
+  pythonPath = [ python3Packages.dnspython tdb ];
```
Comment (SuperSandro2000) [inline pkgs/servers/samba/4.x.nix:149]:
```suggestion
  pythonPath = [ python3.pkgs.dnspython tdb ];
```

File: pkgs/servers/samba/4.x.nix [hunk 4]
```diff
@@ -167,7 +169,10 @@ stdenv.mkDerivation rec {
 
     # Samba does its own shebang patching, but uses build Python
     find "$out/bin" -type f -executable -exec \
-      sed -i '1 s^#!${python3.pythonForBuild}/bin/python.*^#!${python3.interpreter}^' {} \;
+      sed -i '1 s^#!${python3Packages.python.pythonForBuild}/bin/python.*^#!${python3Packages.python.interpreter}^' {} \;
```
Comment (SuperSandro2000) [inline pkgs/servers/samba/4.x.nix:172]:
```suggestion
      sed -i '1 s^#!${python3.pythonForBuild}/bin/python.*^#!${python3.interpreter}^' {} \;
```

**Review by SuperSandro2000 — COMMENTED** [review]
when we use python3 instead of python3Packages we don't need to use python3Packages.python.

**Comment by Artturin** [comment]
> when we use python3 instead of python3Packages we don't need to use python3Packages.python.

eh i think this is nitpicking

**Review by veprbl — APPROVED** [review]
Diff LGTM

**Comment by kwaerner** [comment]
Thank you all for fixing this so quickly! Would it be possible to get this backported to 21.05?

**Pull request merged**